### PR TITLE
C++: Add isBarrier to cpp/cgi-xss

### PR DESCRIPTION
**Pull request by MathiasVP**
This fixes the FP reported [here](https://github.com/github/codeql/issues/5162). The solution is the exact same as the one applied in https://github.com/github/codeql/pull/4915.

Here's a rundown of the alerts I investigated:
- The PR removes all alerts for this query on `timsgit/ipscan` (which are currently all marked with `lgtm[cpp/cgi-xss]` in the comments).
- We keep the 1 result on `Skycrab/Linux-C-Web-Server` (which looks like a TP).
- We keep the 1 result on `stealth/crash` (which looks like a TP).
- We keep the 1 result on `oetiker/rrdtool-1.x` (which looks like a TP).
- We keep the 2 results on `Maxime2/dataparksearch` (which both look like TPs).

(Closes https://github.com/github/codeql-c-analysis-team/issues/234)

### CODE REVIEW

**Comment by rdmarsh2** [comment]
Should the barrier use `ArithmeticType` instead?

**Comment by MathiasVP** [comment]
> Should the barrier use `ArithmeticType` instead?

Good point! I think that might be a good idea both for this PR and as a follow-up to https://github.com/github/codeql/pull/4915.

File: cpp/ql/test/query-tests/Security/CWE/CWE-079/semmle/CgiXss/search.c [hunk 1]
```diff
@@ -37,14 +37,26 @@ void good_server(char* query) {
...
 }
 
+int snprintf(char *, int, const char *, ...);
+
+void good_server2(char* query) {
+  puts("<p>Query results for ");
+  // GOOD: Only an integer is added to the page.
+  int i = 0;
+  snprintf(query, 16, "value=%i", &i);
+  printf("\n<p>%i</p>\n", i);
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Security/CWE/CWE-079/semmle/CgiXss/search.c:47]: Would this have produced a false positive result without the fix in this PR?  Is `snprintf` supposed to be `scanf`???

File: cpp/ql/src/Security/CWE/CWE-079/CgiXss.ql [hunk 2]
```diff
@@ -34,6 +34,10 @@ class Configuration extends TaintTrackingConfiguration {
   override predicate isSink(Element tainted) {
     exists(PrintStdoutCall call | call.getAnArgument() = tainted)
   }
+
+  override predicate isBarrier(Expr e) {
+    super.isBarrier(e) or e.getUnspecifiedType() instanceof IntegralType
+  }
```
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-079/CgiXss.ql:40]: Change LGTM.

File: cpp/ql/test/query-tests/Security/CWE/CWE-079/semmle/CgiXss/search.c (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-079/semmle/CgiXss/search.c:47]: Oops, you're right. Good catch! I think I mixed up my test cases. Indeed, this new test does not produce a false positive on `main` currently. It does if I replace `snprintf` with `sscanf` (which is what I wanted to write). I've fixed this in f5d5460dde1768b82c7ac21bd022d99f8d3e427e.

File: cpp/ql/test/query-tests/Security/CWE/CWE-079/semmle/CgiXss/search.c [hunk 3]
```diff
@@ -37,14 +37,26 @@ void good_server(char* query) {
   puts(do_search(query));
 }
 
+int scanf(const char *, ...);
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Security/CWE/CWE-079/semmle/CgiXss/search.c:40]:
```suggestion
int sscanf(const char *s, const char *format, ...);
```

Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-079/semmle/CgiXss/search.c:40]: Jeez, I need to read what I actually write before I commit it >.< Thanks a lot!

Comment (geoffw0) [reply cpp/ql/test/query-tests/Security/CWE/CWE-079/semmle/CgiXss/search.c]: You were 'saved' by C implicit declarations!

**Review by geoffw0 — APPROVED** [review]
LGTM.

**Comment by troy1993-tcs** [comment]
there the simulation cam-boos is selected to zero ;
therefore potential result Negetive or Positive  cannot be calculated. here by

**Pull request merged**